---
title: "Illustration on Alpha/Beta Diversities"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
rm(list = ls())
library(microbiome)
library(tidyverse)
library(ape)
library(ggpubr)
```

# 1. Load Example Data

```{r, message=FALSE, warning=FALSE, comment=NA}
data(dietswap)
pseq = dietswap
n_taxa = ntaxa(pseq)
n_samp = nsamples(pseq)
# Metadata
meta_data = meta(pseq)
# Taxonomy table
taxonomy = tax_table(pseq)
# Phylogenetic tree
random_tree = rtree(n = n_taxa, rooted = TRUE, tip.label = taxa(pseq))
# plot(random_tree)
pseq = merge_phyloseq(pseq, random_tree)
pseq
# Absolute abundances
otu_absolute = abundances(pseq)
# Relative abundances
otu_relative = abundances(pseq, "compositional")
```

# 2. Alpha diversities

```{r, message=FALSE, warning=FALSE, comment=NA, results='hide'}
alpha_tab = alpha(pseq, index = "all")
# To get PD
library(picante)
alpha_tab$pd = pd(samp = t(otu_absolute), tree = random_tree)$PD
p_df = alpha_tab%>%select(observed, chao1, diversity_shannon, diversity_gini_simpson, pd)
colnames(p_df) = c("Observed Species", "Chao1", "Shannon's Diversity", "Simpson's Diversity", "PD")

library(GGally)
p = ggpairs(p_df, aes(alpha = 0.4))
```

```{r, message=FALSE, warning=FALSE, comment=NA}
p
ggsave("../figures/alpha_corr.jpeg", width=7, height=7, units='in', dpi = 300)
ggsave("../figures/alpha_corr.pdf", width=7, height=7, units='in')
```

# 3. Beta diversities

```{r, message=FALSE, warning=FALSE, comment=NA, results='hide'}
set.seed(1234)
pseq_rarified = rarefy_even_depth(pseq)

# unlist(distanceMethodList)
dist_methods = c("bray", "jaccard", "unifrac", "wunifrac")
plist = vector("list", length(dist_methods))
names(plist) = dist_methods
for(i in dist_methods){
  # Calculate distance matrix
  iDist = distance(pseq_rarified, method = i)
  # Calculate ordination
  iMDS  = ordinate(pseq_rarified, "MDS", distance = iDist)
  ## Make plot
  # Don't carry over previous plot (if error, p will be blank)
  p = NULL
  # Create plot, store as temp variable, p
  p = plot_ordination(pseq_rarified, iMDS, color = "nationality")
  # Save the graphic to file.
  plist[[i]] = p
}

p_df = plist%>%map_dfr(function(x) x$data, .id = "distance")
p_df$distance = recode(p_df$distance, 
                       bray = "Bray-Curtis", 
                       jaccard = "Jaccard", 
                       unifrac = "Unweighted UniFrac", 
                       wunifrac = "Weighted UniFrac")
p_df$nationality = recode(p_df$nationality,
                          AAM = "African American",
                          AFR = "Native African")
p = ggplot(p_df, aes(Axis.1, Axis.2, color=nationality)) + 
  geom_point(size=3, alpha=0.5) + 
  labs(x = "Coordinate 1", y = "Coordinate 2") + 
  scale_color_discrete(name = NULL)+
  facet_wrap(~distance, scales="free") + 
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5),
        strip.background = element_rect(fill = "white"),
        legend.position = "bottom")
```

```{r, message=FALSE, warning=FALSE, comment=NA}
p
ggsave("../figures/beta_pcoa.jpeg", width=6.25, height=5, units='in', dpi = 300)
ggsave("../figures/beta_pcoa.pdf", width=6.25, height=5, units='in')
```

# 4. Rarefaction curve

```{r, message=FALSE, warning=FALSE, comment=NA, results='hide'}
# Calculate alpha diversity
calculate_rarefaction_curves = function(psdata, measures, depths) {
  estimate_rarified_richness = function(psdata, measures, depth) {
    if(max(sample_sums(psdata)) < depth) return()
    
    psdata = prune_samples(sample_sums(psdata) >= depth, psdata)
    rarified_psdata = rarefy_even_depth(psdata, depth, verbose = FALSE)
    alpha_diversity = alpha(rarified_psdata, index = measures)
    molten_alpha_diversity = alpha_diversity%>%rownames_to_column("sample")%>%
      gather(key = "measure", value = "alpha_diversity", -sample)
    return(molten_alpha_diversity)
  }
  
  names(depths) = depths
  rarefaction_curve_data = depths%>%map_dfr(function(x)
    estimate_rarified_richness(psdata = psdata, measures = measures, depth = x), .id = "depth")
  
  return(rarefaction_curve_data)
}

rarefaction_curve_data = calculate_rarefaction_curves(psdata = pseq, 
                                                      measures = c("chao1", "diversity_shannon"), 
                                                      depths = rep(round(seq(1, max(sample_sums(pseq)), length.out = 20)), 
                                                                   each = 10))
rarefaction_curve_data$depth = as.numeric(rarefaction_curve_data$depth)
# Summarize alpha diversity
rarefaction_curve_data_summary = rarefaction_curve_data%>%group_by(depth, sample, measure)%>%
  summarise(alpha_diversity_mean = mean(alpha_diversity), alpha_diversity_sd = sd(alpha_diversity))

# Add sample data
p_df = rarefaction_curve_data_summary%>%left_join(meta_data, by = "sample")

# Plot
p_df$measure = recode(p_df$measure, 
                      chao1 = "Chao1", 
                      diversity_shannon = "Shannon's Diversity")
p_df$nationality = recode(p_df$nationality,
                          AAM = "African American",
                          AFR = "Native African")
p = ggplot(data = p_df,
           aes(x = depth, y = alpha_diversity_mean,
               ymin = alpha_diversity_mean - alpha_diversity_sd,
               ymax = alpha_diversity_mean + alpha_diversity_sd,
               color = nationality,
               group = sample)) + 
  labs(x = "Library Size", y = "Diversity") + 
  scale_color_discrete(name = NULL) +
  geom_line() + geom_pointrange(alpha = 0.5) + 
  facet_wrap(facets = ~ measure, scales = 'free_y') + 
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5),
        strip.background = element_rect(fill = "white"),
        legend.position = "bottom")
```

```{r, message=FALSE, warning=FALSE, comment=NA}
p
ggsave("../figures/alpha_rare.jpeg", width=6.25, height=5, units='in', dpi = 300)
ggsave("../figures/alpha_rare.pdf", width=6.25, height=5, units='in')
```

# Session information

```{r, message=FALSE, warning=FALSE, comment=NA}
sessionInfo()
devtools::session_info()
```
